### ISSUE TITLE
mysql-haskell-simple

### ISSUE THREAD

**Issue by alexbiehl** [alexbiehl/mysql-haskell-simple#1]
@winterland1989

While tinkering on `persistent-mysql-haskell` another idea for a client library came to my mind. The interface is inspired from `hasql` library for `postgresql`.

The idea is to build a composable way to create queries without being too heavy with all those type level stuff. And it should be performant (persistent does lots of conversions between values).

Take a look: 

```
data Book = Book { bookId :: Int, bookTitle :: String, bookAuthor :: String }

bookByAuthorAndTitle :: Query (String, String) Book
bookByAuthorAndTitle = mkQuery sql params result
  where
    sql = "select * from books where book_title=? and book_author=?"

    params :: Param (String, String)
    params = contramap fst (Enc.value string) 
          <> contramap snd (Enc.value string)

    dec :: Row Book 
    dec = Book <$> Dec.value int <*> Dec.value string <*> Dec.value string  

    result :: Result Book
    result = Result.singleRow dec
```

For query params we the `Param` type which can be composed using its `Monoid` instance also its an instance of `Contravariant`.

The resulting rows can be decoded with a `Row`. Rows can be composed using applicative operations.

Now, we want to control the `Result` of a query. Wether its a unit result, or a single row expected or the result als `Vector` or a streaming fold over the result the `Result` type is your friend. 

Internally `Param` and `Row` are implemented very efficiently. Encoding and decoding of `MySQLValue`s is very straight in GHCs core/stg which makes it a quite performant.

What do you think?

**Comment by alexbiehl** [comment]
For example here, a `Param User` values encoding function:

```haskell
data User = User { usrName :: Int, usrAge :: Int, usrBlah :: Int }

{-
params
  :: forall s.
     User
     -> Int -> STVector s MySQLValue -> State# s -> (# State# s, Int #)
params =
  \ (@ s)
    (a8 :: User)
    (i :: Int)
    (v :: STVector s MySQLValue)
    (s1 :: State# s) ->
    case v `cast` ... of _ { MVector dt dt1 dt2 ->
    case i of _ { I# y ->
    case writeArray#
           dt2
           (+# dt y)
           (MY_Int (case a8 of _ { User ds ds1 ds2 -> ds }))
           (s1 `cast` ...)
    of s'# { __DEFAULT ->
    let {
      a9 :: Int#
      a9 = +# y 1 } in
    case writeArray#
           dt2
           (+# dt a9)
           (MY_Int (case a8 of _ { User ds ds1 ds2 -> ds1 }))
           s'#
    of s'#1 { __DEFAULT ->
    let {
      y1 :: Int#
      y1 = +# a9 1 } in
    case writeArray#
           dt2
           (+# dt y1)
           (MY_Int (case a8 of _ { User ds ds1 ds2 -> ds2 }))
           s'#1
    of s'#2 { __DEFAULT ->
    (# s'#2 `cast` ..., I# (+# y1 1) #)
    }
    }
    }
    }
    }
-}
params :: Param User
params =
     contramap usrName (value1 int1)
  <> contramap usrAge (value1 int1)
  <> contramap usrBlah (value1 int1)
```

**Comment by winterland1989** [comment]
I seemed to understand the main idea, so you would want me to export some basic decoders like int, string...right? i'm not familiar with hasql so i can't comment about the detail API design, but if you need any help please let me know ; )

One thing occur to me is that maybe the decoder is derivable from the record(via th or generic), but i can also image there's much work to be done.

**Comment by alexbiehl** [comment]
Currently I create a [`Vector MySQLValue`](https://github.com/alexbiehl/mysql-haskell-simple/blob/master/src/Database/MySQL/Simple/Encoder.hs#L49) for you to serialize. The only thing I'd like to change in `mysql-haskell` for now is the `queryVector` function which should take a `Vector MySQLValue`. Similarly the `query` function should convert the list of query params to a `Vector a` and `COM_STMT_EXECUTE` should take a vector too. What do you think?

**Comment by winterland1989** [comment]
Can you show me how do you construct `Vector MySQLValue`?  I guess i would use list here because in haskell it's hard to construct a `Vector` directly.

**Comment by alexbiehl** [comment]
Take a look at https://github.com/alexbiehl/mysql-haskell-simple/blob/master/src/Database/MySQL/Simple/Encoder.hs#L45

Since `Param` can only be composed through `Applicative` we know the number params statically. So we can create a `Vector` and populate it in one go.

**Comment by winterland1989** [comment]
I see, then i think it best to define vector version here since i have exposed `Query` constructor, you will have to define your own function for to  render params. The version in `mysql-haskell` would should still use list because use vector there simply means adding an extra `fromList`.

**Comment by alexbiehl** [comment]
I am now doing the `toList` myself (https://github.com/alexbiehl/mysql-haskell-simple/blob/master/src/Database/MySQL/Simple/Session.hs#L85). Would you accept a pull request where [COM_STMT_EXECUTE](http://hackage.haskell.org/package/mysql-haskell-0.7.1.0/docs/Database-MySQL-Protocol-Command.html#v:COM_STMT_EXECUTE)  takes the null bitmap explicitly? I can construct this bitmap without going over the data twice.

**Comment by winterland1989** [comment]
Yes, I think i'll accept a patch to move this bitmap to `COM_STMT_EXECUTE`, then `executeStmt` should to be changed accordingly. But let's not add `executeStmtWithBitmap` to `mysql-haskell`, it's too exotic to normal user ; )

**Comment by alexbiehl** [comment]
I wrote the custom function: https://github.com/alexbiehl/mysql-haskell-simple/blob/master/src/Database/MySQL/Simple/Internal.hs#L17

Lets see if I can hack up the benchmark
